### PR TITLE
Make SSL_OP_ALLOW_NO_DHE_KEX actually skip ECDHE

### PR DESCRIPTION
**Pull request by bdonlan**
Currently, SSL_OP_ALLOW_NO_DHE_KEX is ineffective, as on the client side a key
share extension is sent regardless, and on the server side we will always send
a key share if the client sent a key share. What's more, the server will fail
with an internal error if a client key is missing later on in the exchange.

To fix this, we change the client side logic to skip sending a key share
extension if SSL_OP_ALLOW_NO_DHE_KEX is provided, a PSK session is available,
and we are not performing a hello retry. In order for this to work, the PSK
session is now determined when we generate the PSK KEX modes extension, which
comes before the key share extension in the handshake. We also fix the
server-side sanity check to allow a missing ckey if SSL_OP_ALLOW_NO_DHE_KEX is
set, and a PSK session is in use.

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Comment by bdonlan** [comment]
Note that this patch applies cleanly against both master and 1.1.1.

**Comment by bdonlan** [comment]
Ping?

File: ssl/statem/extensions_clnt.c [hunk 1]
```diff
@@ -579,7 +668,7 @@ EXT_RETURN tls_construct_ctos_psk_kex_modes(SSL *s, WPACKET *pkt,
     s->ext.psk_kex_mode = TLSEXT_KEX_MODE_FLAG_KE_DHE;
     if (nodhe)
         s->ext.psk_kex_mode |= TLSEXT_KEX_MODE_FLAG_KE;
-#endif
+#endif  /* OPENSSL_NO_TLS1_3 */
 
     return EXT_RETURN_SENT;
```
Comment (mattcaswell) [inline ssl/statem/extensions_clnt.c]: Looks like there is a bug here in the existing code. `EXT_RETURN_SENT` should only be returned if the extension has actually been added. In the case of `OPENSSL_NO_TLS1_3` it isn't sent so we should return `EXT_RETURN_NOT_SENT`.

File: ssl/statem/extensions_clnt.c [hunk 2]
```diff
@@ -652,6 +741,12 @@ EXT_RETURN tls_construct_ctos_key_share(SSL *s, WPACKET *pkt,
     const uint16_t *pgroups = NULL;
     uint16_t curve_id = 0;
 
+    /* Send an empty key share if we are allowing a non-DHE key exchange method, and are not retrying the ClientHello. */
+    if (s->s3->group_id == 0 && s->psksession != NULL && s->options & SSL_OP_ALLOW_NO_DHE_KEX) {
+        /* Don't add a key share */
```
Comment (mattcaswell) [inline ssl/statem/extensions_clnt.c]: The comment above says send an empty key share, but here we say don't add a key share at all. These are different things.

**Comment by mattcaswell** [comment]
> I am unconvinced by the premise of this PR. SSL_OP_ALLOW_NO_DHE_KEX is about allowing a non (ec)dhe key exchange to occur not about forcing it. This PR would change the meaning of that option. Conceivably we could imagine a new option that forced it, but I'm not sure if that is a good idea.

For the above reasons, I'm closing this PR.

**Pull request closed without merging**